observability/src/main/java/io/grpc/observability/interceptors/LogHelper.java

```diff
@@ -0,0 +1,436 @@
+ * limitations under the License.
+ */
+
+package io.grpc.observability.interceptors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Charsets;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.Duration;
+import com.google.protobuf.util.Timestamps;
+import io.grpc.Attributes;
+import io.grpc.Deadline;
+import io.grpc.Grpc;
+import io.grpc.InternalMetadata;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor.Marshaller;
+import io.grpc.Status;
+import io.grpc.internal.TimeProvider;
+import io.grpc.observability.logging.Sink;
+import io.grpc.observabilitylog.v1.GrpcLogRecord;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.Address;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.EventType;
+import io.grpc.observabilitylog.v1.GrpcLogRecord.LogLevel;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.Inet4Address;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * Helper class for GCP observability logging.
+ */
+class LogHelper {
+  private static final Logger logger = Logger.getLogger(LogHelper.class.getName());
+  // TODO(dnvindhya): Define it in one places(TBD) to make it easily accessible from everywhere
+  static final Metadata.Key<byte[]> STATUS_DETAILS_KEY =
+      Metadata.Key.of(
+          "grpc-status-details-bin",
+          Metadata.BINARY_BYTE_MARSHALLER);
+
+  /**
+   *  Class for proto related helper methods.
+   */
+  static final class LogSinkWriter {
+    private final Sink sink;
+    private final TimeProvider timeProvider;
+
+    LogSinkWriter(Sink sink, TimeProvider timeProvider) {
+      this.sink = sink;
+      this.timeProvider = timeProvider;
+    }
+
+    GrpcLogRecord.Builder createTimestamp() {
+      long nanos = timeProvider.currentTimeNanos();
+      return GrpcLogRecord.newBuilder().setTimestamp(Timestamps.fromNanos(nanos));
+    }
+
+
+    /**
+     * Logs the request header.
+     * Binary logging equivalent of logClientHeader.
+     */
+    void logRequestHeader(
+        long seqId,
+        String serviceName,
+        String methodName,
+        String authority,
+        @Nullable Duration timeout,
+        Metadata metadata,
+        GrpcLogRecord.EventLogger eventLogger,
+        String rpcId,
+        // null on client side
+        @Nullable SocketAddress peerAddress) {
+      checkNotNull(serviceName, "serviceName");
+      checkNotNull(methodName, "methodName");
+      checkNotNull(rpcId, "rpcId");
+      checkArgument(
+          peerAddress == null || eventLogger == GrpcLogRecord.EventLogger.LOGGER_SERVER,
```

Comment:
I would like to see a comment here explaining why peerAddress is null on the client side - because the connection may not be established yet?